File: react-native/ios/MobModule.swift
import ExpoModulesCore

public class MobModule: Module {
  // In-memory registries keyed by string identifiers
  private var signers: [String: RustSigner] = [:]
  private var keychainSigners: [String: KeychainSigner] = [:]
  private var clients: [String: Client] = [:]
  private var sessionManagers: [String: MobSessionManager] = [:]
  private var clientCounter: Int = 0
  private var sessionCounter: Int = 0
  private let transport: HttpTransport = NativeHttpTransport()

  public func definition() -> ModuleDefinition {
    Name("Mob")

    // MARK: - Signer Management

    AsyncFunction("createSignerFromMnemonic") { (mnemonic: String, prefix: String, derivationPath: String?) -> [String: String] in
      let signer = try RustSigner.fromMnemonic(
        mnemonic: mnemonic,
        addressPrefix: prefix,
        derivationPath: derivationPath
      )
      let address = signer.address()
      let pubKeyHex = signer.publicKeyHex()
      self.signers[address] = signer
      return ["address": address, "publicKeyHex": pubKeyHex]
    }

    AsyncFunction("signBytes") { (signerAddress: String, message: [UInt8]) -> [UInt8] in
      guard let signer = self.signers[signerAddress] else {
        throw MobModuleError.signerNotFound(signerAddress)
      }
      let result = try signer.signBytes(message: Data(message))
      return [UInt8](result)
    }

    AsyncFunction("destroySigner") { (address: String) in
      self.signers.removeValue(forKey: address)
    }

    // MARK: - Keychain Signer (iOS Secure Key Storage)

    AsyncFunction("createKeychainSigner") { (mnemonic: String, prefix: String, derivationPath: String?, identifier: String, requireBiometrics: Bool?) -> [String: String] in
      let signer = try KeychainSigner.fromMnemonic(
        mnemonic: mnemonic,
        prefix: prefix,
        derivationPath: derivationPath,
        identifier: identifier,
        requireBiometrics: requireBiometrics ?? false
      )
      let address = signer.address()
      self.keychainSigners[address] = signer
      return [
        "address": address,
        "publicKeyHex": signer.publicKey().map { String(format: "%02x", $0) }.joined(),
      ]
    }

    AsyncFunction("loadKeychainSigner") { (identifier: String, prefix: String, derivationPath: String?) -> [String: String] in
      let signer = try KeychainSigner.load(
        identifier: identifier,
        prefix: prefix,
        derivationPath: derivationPath
      )
      let address = signer.address()
      self.keychainSigners[address] = signer
      return [
        "address": address,
        "publicKeyHex": signer.publicKey().map { String(format: "%02x", $0) }.joined(),
      ]
    }

    AsyncFunction("deleteKeychainSigner") { (identifier: String) in
      // Remove from in-memory registry if loaded
      self.keychainSigners = self.keychainSigners.filter { _, signer in
        return KeychainSigner.exists(identifier: identifier)
      }
      try KeychainSigner.delete(identifier: identifier)
    }

    AsyncFunction("listKeychainSigners") { () -> [String] in
      return try KeychainSigner.listIdentifiers()
    }

    AsyncFunction("keychainSignerExists") { (identifier: String) -> Bool in
      return KeychainSigner.exists(identifier: identifier)
    }

    AsyncFunction("keychainSignBytes") { (signerAddress: String, message: [UInt8]) -> [UInt8] in
      guard let signer = self.keychainSigners[signerAddress] else {
        throw MobModuleError.signerNotFound(signerAddress)
      }
      let result = try signer.signBytes(message: Data(message))
      return [UInt8](result)
    }

    // MARK: - Client Management

    AsyncFunction("createClient") { (config: [String: Any]) -> String in
      let chainConfig = try self.parseChainConfig(config)
      let client = try Client(config: chainConfig, transport: self.transport)
      let clientId = self.nextClientId()
      self.clients[clientId] = client
      return clientId
    }

    AsyncFunction("createClientWithSigner") { (config: [String: Any], signerAddress: String) -> String in
      guard let signer = self.signers[signerAddress] else {
        throw MobModuleError.signerNotFound(signerAddress)
      }
      let chainConfig = try self.parseChainConfig(config)
      let client = try Client.newWithSigner(config: chainConfig, signer: signer, transport: self.transport)
      let clientId = self.nextClientId()
      self.clients[clientId] = client
      return clientId
    }

    AsyncFunction("createClientWithKeychainSigner") { (config: [String: Any], signerAddress: String) -> String in
      guard let signer = self.keychainSigners[signerAddress] else {
        throw MobModuleError.signerNotFound(signerAddress)
      }
      let chainConfig = try self.parseChainConfig(config)
      let client = try Client.newWithCryptoSigner(config: chainConfig, signer: signer, transport: self.transport)
      let clientId = self.nextClientId()
      self.clients[clientId] = client
      return clientId
    }

    AsyncFunction("destroyClient") { (clientId: String) in
      self.clients.removeValue(forKey: clientId)
    }

    // MARK: - Read Queries

    AsyncFunction("getAccount") { (clientId: String, address: String) -> [String: Any] in
      let client = try self.requireClient(clientId)
      let info = try client.getAccount(address: address)
      return [
        "address": info.address,
        "accountNumber": info.accountNumber,
        "sequence": info.sequence,
        "pubKey": info.pubKey as Any,
      ]
    }

    AsyncFunction("getBalance") { (clientId: String, address: String, denom: String) -> [String: String] in
      let client = try self.requireClient(clientId)
      let coin = try client.getBalance(address: address, denom: denom)
      return ["denom": coin.denom, "amount": coin.amount]
    }

    AsyncFunction("getAllBalances") { (clientId: String, address: String) -> [[String: String]] in
      let client = try self.requireClient(clientId)
      let coins = try client.getAllBalances(address: address)
      return coins.map { ["denom": $0.denom, "amount": $0.amount] }
    }

    AsyncFunction("getHeight") { (clientId: String) -> UInt64 in
      let client = try self.requireClient(clientId)
      return try client.getHeight()
    }

    AsyncFunction("getTx") { (clientId: String, hash: String) -> [String: Any] in
      let client = try self.requireClient(clientId)
      let tx = try client.getTx(hash: hash)
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("hasGrants") { (clientId: String, granter: String, grantee: String) -> Bool in
      let client = try self.requireClient(clientId)
      return try client.hasGrants(granter: granter, grantee: grantee)
    }

    AsyncFunction("queryContractSmart") { (clientId: String, contractAddress: String, queryMsg: [UInt8]) -> [UInt8] in
      let client = try self.requireClient(clientId)
      let result = try client.queryContractSmart(contractAddress: contractAddress, queryMsg: Data(queryMsg))
      return [UInt8](result)
    }

    // MARK: - Transactions

    AsyncFunction("send") { (clientId: String, toAddress: String, amount: [[String: String]], memo: String?) -> [String: Any] in
      let client = try self.requireClient(clientId)
      let coins = try self.parseCoins(amount)
      let tx = try client.send(toAddress: toAddress, amount: coins, memo: memo)
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("executeContract") { (clientId: String, contractAddress: String, msg: [UInt8], funds: [[String: String]], memo: String?, gasLimit: UInt64?) -> [String: Any] in
      let client = try self.requireClient(clientId)
      let fundCoins = try self.parseCoins(funds)
      let tx = try client.executeContract(
        contractAddress: contractAddress,
        msg: Data(msg),
        funds: fundCoins,
        memo: memo,
        gasLimit: gasLimit
      )
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("signAndBroadcastMulti") { (clientId: String, messages: [[String: Any]], memo: String?, gasLimit: UInt64?) -> [String: Any] in
      let client = try self.requireClient(clientId)
      let msgs = try self.parseMessages(messages)
      let tx = try client.signAndBroadcastMulti(messages: msgs, memo: memo, gasLimit: gasLimit)
      return self.serializeTxResponse(tx)
    }

    // MARK: - Session Manager

    AsyncFunction("createSessionManager") { (addressPrefix: String) -> String in
      let mgr = MobSessionManager(addressPrefix: addressPrefix)
      let sessionId = self.nextSessionId()
      self.sessionManagers[sessionId] = mgr
      return sessionId
    }

    AsyncFunction("sessionGenerateKey") { (sessionId: String) -> [String: String] in
      let mgr = try self.requireSessionManager(sessionId)
      let info = try mgr.generateSessionKey()
      return ["address": info.address, "publicKeyHex": info.publicKeyHex]
    }

    AsyncFunction("sessionActivate") { (sessionId: String, granter: String, grantee: String, createdAt: UInt64, expiresAt: UInt64, description: String?, config: [String: Any]) in
      let mgr = try self.requireSessionManager(sessionId)
      let metadata = SessionMetadata(
        granter: granter,
        grantee: grantee,
        createdAt: createdAt,
        expiresAt: expiresAt,
        description: description
      )
      let chainConfig = try self.parseChainConfig(config)
      try mgr.activate(metadata: metadata, config: chainConfig, transport: self.transport)
    }

    AsyncFunction("sessionExport") { (sessionId: String) -> [UInt8] in
      let mgr = try self.requireSessionManager(sessionId)
      let result = try mgr.exportSession()
      return [UInt8](result)
    }

    AsyncFunction("sessionRestore") { (data: [UInt8], config: [String: Any]) -> String in
      let chainConfig = try self.parseChainConfig(config)
      let mgr = try MobSessionManager.restore(data: Data(data), config: chainConfig, transport: self.transport)
      let sessionId = self.nextSessionId()
      self.sessionManagers[sessionId] = mgr
      return sessionId
    }

    AsyncFunction("sessionDeactivate") { (sessionId: String) in
      if let mgr = self.sessionManagers[sessionId] {
        mgr.deactivate()
      }
      self.sessionManagers.removeValue(forKey: sessionId)
    }

    AsyncFunction("sessionIsActive") { (sessionId: String) -> Bool in
      guard let mgr = self.sessionManagers[sessionId] else { return false }
      return mgr.isActive()
    }

    AsyncFunction("sessionGranterAddress") { (sessionId: String) -> String? in
      return self.sessionManagers[sessionId]?.granterAddress()
    }

    AsyncFunction("sessionGranteeAddress") { (sessionId: String) -> String? in
      return self.sessionManagers[sessionId]?.granteeAddress()
    }

    AsyncFunction("sessionSignBytes") { (sessionId: String, message: [UInt8]) -> [UInt8] in
      let mgr = try self.requireSessionManager(sessionId)
      let result = try mgr.signBytes(message: Data(message))
      return [UInt8](result)
    }

    // Session-scoped transactions
    AsyncFunction("sessionSend") { (sessionId: String, toAddress: String, amount: [[String: String]], memo: String?) -> [String: Any] in
      let mgr = try self.requireSessionManager(sessionId)
      let client = try mgr.client()
      let coins = try self.parseCoins(amount)
      let tx = try client.send(toAddress: toAddress, amount: coins, memo: memo)
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("sessionExecuteContract") { (sessionId: String, contractAddress: String, msg: [UInt8], funds: [[String: String]], memo: String?, gasLimit: UInt64?) -> [String: Any] in
      let mgr = try self.requireSessionManager(sessionId)
      let client = try mgr.client()
      let fundCoins = try self.parseCoins(funds)
      let tx = try client.executeContract(
        contractAddress: contractAddress,
        msg: Data(msg),
        funds: fundCoins,
        memo: memo,
        gasLimit: gasLimit
      )
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("sessionSignAndBroadcastMulti") { (sessionId: String, messages: [[String: Any]], memo: String?, gasLimit: UInt64?) -> [String: Any] in
      let mgr = try self.requireSessionManager(sessionId)
      let client = try mgr.client()
      let msgs = try self.parseMessages(messages)
      let tx = try client.signAndBroadcastMulti(messages: msgs, memo: memo, gasLimit: gasLimit)
      return self.serializeTxResponse(tx)
    }

    AsyncFunction("sessionQueryContractSmart") { (sessionId: String, contractAddress: String, queryMsg: [UInt8]) -> [UInt8] in
      let mgr = try self.requireSessionManager(sessionId)
      let client = try mgr.client()
      let result = try client.queryContractSmart(contractAddress: contractAddress, queryMsg: Data(queryMsg))
      return [UInt8](result)
    }
  }

  // MARK: - Helpers

  private func nextClientId() -> String {
    clientCounter += 1
    return "client_\(clientCounter)"
  }

  private func nextSessionId() -> String {
    sessionCounter += 1
    return "session_\(sessionCounter)"
  }

  private func requireClient(_ clientId: String) throws -> Client {
    guard let client = clients[clientId] else {
      throw MobModuleError.clientNotFound(clientId)
    }
    return client
  }

  private func requireSessionManager(_ sessionId: String) throws -> MobSessionManager {
    guard let mgr = sessionManagers[sessionId] else {
      throw MobModuleError.sessionNotFound(sessionId)
    }
    return mgr
  }

  private func parseChainConfig(_ dict: [String: Any]) throws -> ChainConfig {
    guard let chainId = dict["chainId"] as? String,
          let rpcEndpoint = dict["rpcEndpoint"] as? String,
          let addressPrefix = dict["addressPrefix"] as? String else {
      throw MobModuleError.invalidInput("Missing required ChainConfig fields: chainId, rpcEndpoint, addressPrefix")
    }
    let config = ChainConfig(
      chainId: chainId,
      rpcEndpoint: rpcEndpoint,
      grpcEndpoint: dict["grpcEndpoint"] as? String,
      addressPrefix: addressPrefix,
      coinType: (dict["coinType"] as? NSNumber)?.uint32Value ?? 118,
      gasPrice: dict["gasPrice"] as? String ?? "0.025",
      feeGranter: dict["feeGranter"] as? String
    )
    return config
  }

  private func parseCoins(_ dicts: [[String: String]]) throws -> [Coin] {
    return try dicts.map { dict in
      guard let denom = dict["denom"], let amount = dict["amount"] else {
        throw MobModuleError.invalidInput("Coin requires 'denom' and 'amount'")
      }
      return Coin(denom: denom, amount: amount)
    }
  }

  private func parseMessages(_ dicts: [[String: Any]]) throws -> [Message] {
    return try dicts.map { dict in
      guard let typeUrl = dict["typeUrl"] as? String,
            let value = dict["value"] as? [UInt8] else {
        throw MobModuleError.invalidInput("Message requires 'typeUrl' and 'value'")
      }
      return Message(typeUrl: typeUrl, value: Data(value))
    }
  }

  private func serializeTxResponse(_ tx: TxResponse) -> [String: Any] {
    return [
      "txhash": tx.txhash,
      "code": tx.code,
      "rawLog": tx.rawLog,
      "gasWanted": tx.gasWanted,
      "gasUsed": tx.gasUsed,
      "height": tx.height,
    ]
  }
}

// MARK: - Errors

enum MobModuleError: Error, LocalizedError {
  case signerNotFound(String)
  case clientNotFound(String)
  case sessionNotFound(String)
  case invalidInput(String)

  var errorDescription: String? {
    switch self {
    case .signerNotFound(let address):
      return "Signer not found for address: \(address)"
    case .clientNotFound(let id):
      return "Client not found: \(id)"
    case .sessionNotFound(let id):
      return "Session manager not found: \(id)"
    case .invalidInput(let msg):
      return "Invalid input: \(msg)"
    }
  }
}

File: react-native/android/src/main/java/com/burnt/mob/MobModule.kt
package com.burnt.mob

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.mob.Client
import uniffi.mob.ChainConfig
import uniffi.mob.Coin
import uniffi.mob.Message
import uniffi.mob.MobSessionManager
import uniffi.mob.RustSigner
import uniffi.mob.SessionMetadata
import com.burnt.mob.NativeHttpTransport

class MobModule : Module() {
    // In-memory registries
    private val signers = mutableMapOf<String, RustSigner>()
    private val clients = mutableMapOf<String, Client>()
    private val sessionManagers = mutableMapOf<String, MobSessionManager>()
    private val transport = NativeHttpTransport()
    private var clientCounter = 0
    private var sessionCounter = 0

    override fun definition() = ModuleDefinition {
        Name("Mob")

        // MARK: Signer Management

        AsyncFunction("createSignerFromMnemonic") { mnemonic: String, prefix: String, derivationPath: String? ->
            val signer = RustSigner.fromMnemonic(mnemonic, prefix, derivationPath)
            val address = signer.address()
            val pubKeyHex = signer.publicKeyHex()
            signers[address] = signer
            mapOf("address" to address, "publicKeyHex" to pubKeyHex)
        }

        AsyncFunction("createSignerFromPrivateKey") { privateKeyHex: String, prefix: String ->
            val keyBytes = privateKeyHex.hexToByteArray()
            val signer = RustSigner.fromPrivateKey(keyBytes.toList().map { it.toUByte() }, prefix)
            val address = signer.address()
            val pubKeyHex = signer.publicKeyHex()
            signers[address] = signer
            mapOf("address" to address, "publicKeyHex" to pubKeyHex)
        }

        AsyncFunction("signBytes") { signerAddress: String, message: List<Int> ->
            val signer = signers[signerAddress]
                ?: throw IllegalArgumentException("Signer not found: $signerAddress")
            val msgBytes = message.map { it.toUByte() }
            signer.signBytes(msgBytes).map { it.toInt() }
        }

        AsyncFunction("destroySigner") { address: String ->
            signers.remove(address)
        }

        // MARK: Client Management

        AsyncFunction("createClient") { config: Map<String, Any?> ->
            val chainConfig = parseChainConfig(config)
            val client = Client(chainConfig, transport)
            val clientId = nextClientId()
            clients[clientId] = client
            clientId
        }

        AsyncFunction("createClientWithSigner") { config: Map<String, Any?>, signerAddress: String ->
            val signer = signers[signerAddress]
                ?: throw IllegalArgumentException("Signer not found: $signerAddress")
            val chainConfig = parseChainConfig(config)
            val client = Client.newWithSigner(chainConfig, signer, transport)
            val clientId = nextClientId()
            clients[clientId] = client
            clientId
        }

        AsyncFunction("destroyClient") { clientId: String ->
            clients.remove(clientId)
        }

        // MARK: Read Queries

        AsyncFunction("getAccount") { clientId: String, address: String ->
            val client = requireClient(clientId)
            val info = client.getAccount(address)
            mapOf(
                "address" to info.address,
                "accountNumber" to info.accountNumber,
                "sequence" to info.sequence,
                "pubKey" to info.pubKey
            )
        }

        AsyncFunction("getBalance") { clientId: String, address: String, denom: String ->
            val client = requireClient(clientId)
            val coin = client.getBalance(address, denom)
            mapOf("denom" to coin.denom, "amount" to coin.amount)
        }

        AsyncFunction("getAllBalances") { clientId: String, address: String ->
            val client = requireClient(clientId)
            client.getAllBalances(address).map {
                mapOf("denom" to it.denom, "amount" to it.amount)
            }
        }

        AsyncFunction("getHeight") { clientId: String ->
            val client = requireClient(clientId)
            client.getHeight()
        }

        AsyncFunction("getTx") { clientId: String, hash: String ->
            val client = requireClient(clientId)
            serializeTxResponse(client.getTx(hash))
        }

        AsyncFunction("hasGrants") { clientId: String, granter: String, grantee: String ->
            val client = requireClient(clientId)
            client.hasGrants(granter, grantee)
        }

        AsyncFunction("queryContractSmart") { clientId: String, contractAddress: String, queryMsg: List<Int> ->
            val client = requireClient(clientId)
            val msgBytes = queryMsg.map { it.toUByte() }
            client.queryContractSmart(contractAddress, msgBytes).map { it.toInt() }
        }

        // MARK: Transactions

        AsyncFunction("send") { clientId: String, toAddress: String, amount: List<Map<String, String>>, memo: String? ->
            val client = requireClient(clientId)
            val coins = parseCoins(amount)
            serializeTxResponse(client.send(toAddress, coins, memo))
        }

        AsyncFunction("executeContract") { clientId: String, contractAddress: String, msg: List<Int>, funds: List<Map<String, String>>, memo: String?, gasLimit: Long? ->
            val client = requireClient(clientId)
            val msgBytes = msg.map { it.toUByte() }
            val fundCoins = parseCoins(funds)
            val gl = gasLimit?.toULong()
            serializeTxResponse(
                client.executeContract(contractAddress, msgBytes, fundCoins, memo, gl)
            )
        }

        AsyncFunction("signAndBroadcastMulti") { clientId: String, messages: List<Map<String, Any?>>, memo: String?, gasLimit: Long? ->
            val client = requireClient(clientId)
            val msgs = parseMessages(messages)
            val gl = gasLimit?.toULong()
            serializeTxResponse(client.signAndBroadcastMulti(msgs, memo, gl))
        }

        // MARK: Session Manager

        AsyncFunction("createSessionManager") { addressPrefix: String ->
            val mgr = MobSessionManager(addressPrefix)
            val sessionId = nextSessionId()
            sessionManagers[sessionId] = mgr
            sessionId
        }

        AsyncFunction("sessionGenerateKey") { sessionId: String ->
            val mgr = requireSessionManager(sessionId)
            val info = mgr.generateSessionKey()
            mapOf("address" to info.address, "publicKeyHex" to info.publicKeyHex)
        }

        AsyncFunction("sessionActivate") { sessionId: String, granter: String, grantee: String, createdAt: Long, expiresAt: Long, description: String?, config: Map<String, Any?> ->
            val mgr = requireSessionManager(sessionId)
            val metadata = SessionMetadata(
                granter = granter,
                grantee = grantee,
                createdAt = createdAt.toULong(),
                expiresAt = expiresAt.toULong(),
                description = description
            )
            val chainConfig = parseChainConfig(config)
            mgr.activate(metadata, chainConfig)
        }

        AsyncFunction("sessionExport") { sessionId: String ->
            val mgr = requireSessionManager(sessionId)
            mgr.exportSession().map { it.toInt() }
        }

        AsyncFunction("sessionRestore") { data: List<Int>, config: Map<String, Any?> ->
            val chainConfig = parseChainConfig(config)
            val dataBytes = data.map { it.toUByte() }
            val mgr = MobSessionManager.restore(dataBytes, chainConfig)
            val sessionId = nextSessionId()
            sessionManagers[sessionId] = mgr
            sessionId
        }

        AsyncFunction("sessionDeactivate") { sessionId: String ->
            sessionManagers[sessionId]?.deactivate()
            sessionManagers.remove(sessionId)
        }

        AsyncFunction("sessionIsActive") { sessionId: String ->
            sessionManagers[sessionId]?.isActive() ?: false
        }

        AsyncFunction("sessionGranterAddress") { sessionId: String ->
            sessionManagers[sessionId]?.granterAddress()
        }

        AsyncFunction("sessionGranteeAddress") { sessionId: String ->
            sessionManagers[sessionId]?.granteeAddress()
        }

        AsyncFunction("sessionSignBytes") { sessionId: String, message: List<Int> ->
            val mgr = requireSessionManager(sessionId)
            val msgBytes = message.map { it.toUByte() }
            mgr.signBytes(msgBytes).map { it.toInt() }
        }

        AsyncFunction("sessionSend") { sessionId: String, toAddress: String, amount: List<Map<String, String>>, memo: String? ->
            val mgr = requireSessionManager(sessionId)
            val client = mgr.client()
            val coins = parseCoins(amount)
            serializeTxResponse(client.send(toAddress, coins, memo))
        }

        AsyncFunction("sessionExecuteContract") { sessionId: String, contractAddress: String, msg: List<Int>, funds: List<Map<String, String>>, memo: String?, gasLimit: Long? ->
            val mgr = requireSessionManager(sessionId)
            val client = mgr.client()
            val msgBytes = msg.map { it.toUByte() }
            val fundCoins = parseCoins(funds)
            val gl = gasLimit?.toULong()
            serializeTxResponse(
                client.executeContract(contractAddress, msgBytes, fundCoins, memo, gl)
            )
        }

        AsyncFunction("sessionSignAndBroadcastMulti") { sessionId: String, messages: List<Map<String, Any?>>, memo: String?, gasLimit: Long? ->
            val mgr = requireSessionManager(sessionId)
            val client = mgr.client()
            val msgs = parseMessages(messages)
            val gl = gasLimit?.toULong()
            serializeTxResponse(client.signAndBroadcastMulti(msgs, memo, gl))
        }

        AsyncFunction("sessionQueryContractSmart") { sessionId: String, contractAddress: String, queryMsg: List<Int> ->
            val mgr = requireSessionManager(sessionId)
            val client = mgr.client()
            val msgBytes = queryMsg.map { it.toUByte() }
            client.queryContractSmart(contractAddress, msgBytes).map { it.toInt() }
        }
    }

    // MARK: Helpers

    private fun nextClientId(): String {
        clientCounter++
        return "client_$clientCounter"
    }

    private fun nextSessionId(): String {
        sessionCounter++
        return "session_$sessionCounter"
    }

    private fun requireClient(clientId: String): Client {
        return clients[clientId]
            ?: throw IllegalArgumentException("Client not found: $clientId")
    }

    private fun requireSessionManager(sessionId: String): MobSessionManager {
        return sessionManagers[sessionId]
            ?: throw IllegalArgumentException("Session manager not found: $sessionId")
    }

    private fun parseChainConfig(config: Map<String, Any?>): ChainConfig {
        val chainId = config["chainId"] as? String
            ?: throw IllegalArgumentException("Missing chainId")
        val rpcEndpoint = config["rpcEndpoint"] as? String
            ?: throw IllegalArgumentException("Missing rpcEndpoint")
        val addressPrefix = config["addressPrefix"] as? String
            ?: throw IllegalArgumentException("Missing addressPrefix")

        return ChainConfig(
            chainId = chainId,
            rpcEndpoint = rpcEndpoint,
            grpcEndpoint = config["grpcEndpoint"] as? String,
            addressPrefix = addressPrefix,
            coinType = (config["coinType"] as? Number)?.toInt()?.toUInt() ?: 118u,
            gasPrice = config["gasPrice"] as? String ?: "0.025",
            feeGranter = config["feeGranter"] as? String
        )
    }

    private fun parseCoins(dicts: List<Map<String, String>>): List<Coin> {
        return dicts.map { dict ->
            Coin(
                denom = dict["denom"] ?: throw IllegalArgumentException("Coin missing 'denom'"),
                amount = dict["amount"] ?: throw IllegalArgumentException("Coin missing 'amount'")
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseMessages(dicts: List<Map<String, Any?>>): List<Message> {
        return dicts.map { dict ->
            val typeUrl = dict["typeUrl"] as? String
                ?: throw IllegalArgumentException("Message missing 'typeUrl'")
            val value = (dict["value"] as? List<Int>)?.map { it.toUByte() }
                ?: throw IllegalArgumentException("Message missing 'value'")
            Message(typeUrl = typeUrl, value = value)
        }
    }

    private fun serializeTxResponse(tx: uniffi.mob.TxResponse): Map<String, Any> {
        return mapOf(
            "txhash" to tx.txhash,
            "code" to tx.code,
            "rawLog" to tx.rawLog,
            "gasWanted" to tx.gasWanted,
            "gasUsed" to tx.gasUsed,
            "height" to tx.height
        )
    }
}

private fun String.hexToByteArray(): ByteArray {
    val hex = if (startsWith("0x")) substring(2) else this
    require(hex.length % 2 == 0) { "Hex string must have even length" }
    return ByteArray(hex.length / 2) { i ->
        hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
    }
}
